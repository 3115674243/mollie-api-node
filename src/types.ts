import createMollieClient from './createMollieClient';
export default createMollieClient;

export * from './createMollieClient';

export { default as List } from './data/list/List';

export { default as Capture } from './data/payments/captures/Capture';
import {
  GetParameters as CapturesGetParameters,
  ListParameters as CapturesListParameters
} from './resources/payments/captures/parameters';
export { CapturesGetParameters, CapturesListParameters };

export { default as Chargeback } from './data/chargebacks/Chargeback';
import {
  ListParameters as ChargebacksListParameters
} from './resources/chargebacks/parameters';
export { ChargebacksListParameters };

export { default as Customer } from './data/customers/Customer';
import {
  CreateParameters as CustomerCreateParams,
  GetParameters as CustomerGetParams,
  ListParameters as CustomersListParams,
  UpdateParameters as CustomerUpdateParams,
  DeleteParameters as CustomerDeleteParams,
} from './resources/customers/parameters';
export { CustomerCreateParams, CustomerGetParams, CustomersListParams, CustomerUpdateParams, CustomerDeleteParams };

export { default as Mandate } from './data/customers/mandates/Mandate';
import { CreateParameters as MandateCreateParams, GetParameters as MandateGetParams, ListParameters as MandatesListParams, RevokeParameters as MandateRevokeParams } from './resources/customers/mandates/parameters';
export { MandateCreateParams, MandateGetParams, MandatesListParams, MandateRevokeParams };

export { default as Method } from './data/methods/Method';
import { GetParameters as MethodGetParams, ListParameters as MethodsListParams } from './resources/methods/parameters';
export { MethodGetParams, MethodsListParams };

export { default as Order } from './data/orders/Order';
import {
  CreateParameters as OrderCreateParams,
  GetParameters as OrderGetParams,
  ListParameters as OrdersListParams,
  UpdateParameters as OrderUpdateParams,
  CancelParameters as OrderCancelParams,
} from './resources/orders/parameters';
export { OrderCreateParams, OrderGetParams, OrdersListParams, OrderUpdateParams, OrderCancelParams };

export { default as OrderLine } from './data/orders/orderlines/OrderLine';
import { UpdateParameters as OrderLineUpdateParams, CancelParameters as OrderLineCancelParams } from './resources/orders/orderlines/parameters';
export { OrderLineUpdateParams, OrderLineCancelParams };

export { default as Payment } from './data/payments/Payment';
import { CreateParameters as PaymentCreateParams, GetParameters as PaymentGetParams, ListParameters as PaymentsListParams, CancelParameters as PaymentCancelParams } from './resources/payments/parameters';
import { CreateParameters as CustomerPaymentCreateParams, ListParameters as CustomerPaymentsListParams } from './resources/customers/payments/parameters';
import { CreateParameters as OrderPaymentCreateParams } from './resources/payments/orders/parameters';
export {
  PaymentCreateParams,
  PaymentGetParams,
  PaymentsListParams,
  PaymentCancelParams,
  CustomerPaymentCreateParams,
  CustomerPaymentsListParams,
  OrderPaymentCreateParams,
};

export { default as Refund } from './data/refunds/Refund';
import { ListParameters as RefundsListParams } from './resources/refunds/parameters';
import {
  CreateParameters as PaymentRefundCreateParams,
  GetParameters as PaymentRefundGetParams,
  ListParameters as PaymentRefundsListParams,
  CancelParameters as PaymentRefundCancelParams,
} from './resources/payments/refunds/parameters';
export { RefundsListParams, PaymentRefundCreateParams, PaymentRefundGetParams, PaymentRefundsListParams, PaymentRefundCancelParams };

export { default as Shipment } from './data/orders/shipments/Shipment';
import { CreateParameters as ShipmentCreateParams, GetParameters as ShipmentGetParams, ListParameters as ShipmentsListParams, UpdateParameters as ShipmentUpdateParams } from './resources/orders/shipments/parameters';
export { ShipmentCreateParams, ShipmentGetParams, ShipmentsListParams, ShipmentUpdateParams };

export { default as Subscription } from './data/subscription/Subscription';
import {
  CreateParameters as SubscriptionCreateParams,
  GetParameters as SubscriptionGetParams,
  ListParameters as SubscriptionsListParams,
  UpdateParameters as SubscriptionUpdateParams,
  CancelParameters as SubscriptionCancelParams,
} from './resources/customers/subscriptions/parameters';
export { SubscriptionCreateParams, SubscriptionGetParams, SubscriptionsListParams, SubscriptionUpdateParams, SubscriptionCancelParams };

export { CardAudience, CardFailureReason, CardLabel, FeeRegion } from './data/global';
export { Issuer } from './data/Issuer';
export { PaymentInclude } from './data/payments/data';
