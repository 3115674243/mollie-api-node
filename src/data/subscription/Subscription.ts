import { SubscriptionData } from './data';
import Seal from '../../types/Seal';
import subscriptionHelpers from './helpers';

type Subscription = Seal<SubscriptionData, typeof subscriptionHelpers>;

export default Subscription;

export function injectPrototypes(input: SubscriptionData): Subscription {
  return Object.assign(Object.create(subscriptionHelpers), input);
}
