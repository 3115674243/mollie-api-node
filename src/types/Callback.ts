type Callback<R> = (error: Error | null, result: R) => void;

export default Callback;
