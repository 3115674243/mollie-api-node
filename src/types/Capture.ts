declare namespace Mollie {
  interface ICapture {
    resource: string;
    id: string;
    mode: ApiMode;
    amount: IAmount;
    settlementAmount: IAmount;
    paymentId: string;
    shipmentId?: string;
    settlementId?: string;
    createdAt: string;
    _links: ILinks;
  }

  namespace Params {
    interface Create {
      // Access token parameters
      testmode?: false;
    }
  }
}
