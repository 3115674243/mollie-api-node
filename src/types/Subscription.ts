declare namespace Mollie {
  interface ISubscription {
    resource: string;
    id: string;
    mode: ApiMode;
    status: Subscription.Status;
    amount: IAmount;
    times: number;
    timesRemaining: number;
    interval: string;
    startDate: string;
    nextPaymentDate?: string;
    description: string;
    method: string;
    mandateId?: string;
    createdAt?: string;
    canceledAt: string;
    webhookUrl: string;
    metadata: any;
    _links: ILinks;

    // Access token parameters
    testmode?: boolean;
  }

  namespace Subscription {
    enum Status {
      pending = 'pending',
      active = 'active',
      canceled = 'canceled',
      suspended = 'suspended',
      completed = 'completed',
    }

    namespace Params {
      interface ICreate {
        customerId: string;
        amount: IAmount;
        times?: number;
        interval: string;
        startDate?: string;
        description: string;
        method: Mandate.Method | null;
        mandateId?: string;
        webhookUrl?: string;
        metadata?: string;

        // Access token parameters
        profileId?: string;
        testmode?: boolean;
      }

      interface IGet {
        customerId: string;

        // Access token parameters
        testmode?: boolean;
      }

      interface IUpdate extends ISubscription {
        customerId: string;

        // Access token parameters
        testmode?: boolean;
      }

      interface ICancel {
        customerId: string;

        // Access token parameters
        testmode?: boolean;
      }

      interface IList {
        customerId: string;

        from?: string;
        limit?: number;

        // Access token parameters
        profileId?: string;
        testmode?: boolean;
      }
    }

    namespace Callback {
      type Create = (error: any, subscription: ISubscription) => void;
      type Get = (error: any, subscription: ISubscription) => void;
      type List = (error: any, subscriptions: Mollie.List<ISubscription>) => void;
      type Update = (error: any, subscription: ISubscription) => void;
      type Cancel = (error: any, success: boolean) => void;
    }

    namespace Payment {
      namespace Params {
        interface IList {
          customerId: string;
          subscriptionId: string;

          from?: string;
          limit?: number;

          // Access token parameters
          profileId?: string;
          testmode?: boolean;
        }
      }

      namespace Callback {
        type List = (error: any, payments: Mollie.List<IPayment>) => void;
      }
    }
  }
}
