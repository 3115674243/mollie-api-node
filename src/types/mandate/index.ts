import { ApiMode, CardLabel, ILinks, IUrl } from '../global';

/**
 * Mandate Response object
 *
 * @see https://docs.mollie.com/reference/v2/mandates-api/get-mandate
 */
export interface IMandate {
  resource: string;
  id: string;
  mode: ApiMode;
  status: MandateStatus;
  method: MandateMethod;
  details: IMandateDetails;
  mandateReference: string;
  signatureDate: string;
  createdAt: string;
  _links: IMandateLinks;

  // Access token parameters
  testmode?: boolean;
}

export interface IMandateLinks extends ILinks {
  customer: IUrl;
}

export type IMandateDetails = IMandateDetailsCreditCard | IMandateDetailsDirectDebit;

export interface IMandateDetailsDirectDebit {
  consumerName: string;
  consumerAccount: string;
  consumerBic: string;
}

export interface IMandateDetailsCreditCard {
  cardHolder: string;
  cardNumber: string;
  cardLabel: CardLabel;
  cardFingerprint: string;
  cardExpireDate: string;
}

export enum MandateMethod {
  directdebit = 'directdebit',
  creditcard = 'creditcard',
}

export enum MandateStatus {
  valid = 'valid',
  invalid = 'invalid',
  pending = 'pending',
}
