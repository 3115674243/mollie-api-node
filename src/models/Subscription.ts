import Model from '../model';
import { ISubscription, SubscriptionStatus } from '../types/subscription';

/**
 * The `Subscription` model
 *
 * {@link ISubscription}
 */
export default class Subscription extends Model implements ISubscription {
  public static resourcePrefix = 'sub_';

  public resource = 'subscription';
  public id = null;
  public mode = null;
  public createdAt = null;
  public status = null;
  public amount = {
    currency: null,
    value: null,
  };
  public times = null;
  public interval = null;
  public startDate = null;
  public description = null;
  public method = null;
  public canceledAt = null;
  public webhookUrl = null;
  public timesRemaining = null;
  public metadata = null;
  public _links = {
    self: null,
    documentation: null,
    customer: null,
  };
  public mandateId = null;
  public nextPaymentDate = null;

  // Access token parameters
  public testmode?: boolean;

  /**
   * Subscription constructor
   *
   * @public ✓ This constructor is part of the public API
   */
  public constructor(props?: Partial<ISubscription>) {
    super();

    Object.assign(this, props);
  }

  /**
   * Get the webhook url
   *
   * @public ✓ This method is part of the public API
   */
  public getWebhookUrl(): string {
    return this.webhookUrl;
  }

  /**
   * If the subscription is active
   *
   * @public ✓ This method is part of the public API
   */
  public isActive(): boolean {
    return this.status === SubscriptionStatus.active;
  }

  /**
   * If the subscription is pending
   *
   * @public ✓ This method is part of the public API
   */
  public isPending(): boolean {
    return this.status === SubscriptionStatus.pending;
  }

  /**
   * If the subscription is completed
   *
   * @public ✓ This method is part of the public API
   */
  public isCompleted(): boolean {
    return this.status === SubscriptionStatus.completed;
  }

  /**
   * If the subscription is suspended
   *
   * @public ✓ This method is part of the public API
   */
  public isSuspended(): boolean {
    return this.status === SubscriptionStatus.suspended;
  }

  /**
   * If the subscription is canceled
   *
   * @public ✓ This method is part of the public API
   */
  public isCanceled(): boolean {
    return SubscriptionStatus.canceled == this.status;
  }
}
