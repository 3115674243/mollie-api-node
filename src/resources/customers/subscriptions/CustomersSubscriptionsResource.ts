import { CancelParameters, CreateParameters, GetParameters, ListParameters, UpdateParameters } from './parameters';
import { SubscriptionData } from '../../../data/subscription/data';
import ApiError from '../../../errors/ApiError';
import Callback from '../../../types/Callback';
import List from '../../../data/list/List';
import ParentedResource from '../../ParentedResource';
import Subscription, { injectPrototypes } from '../../../data/subscription/Subscription';
import checkId from '../../../plumbing/checkId';
import renege from '../../../plumbing/renege';

/**
 * The `customers_subscriptions` resource.
 *
 * @since 1.3.2
 */
export default class CustomersSubscriptionsResource extends ParentedResource<SubscriptionData, Subscription> {
  protected getResourceUrl(customerId: string): string {
    return `customers/${customerId}/subscriptions`;
  }

  protected injectPrototypes = injectPrototypes;

  /**
   * Delete a customer subscription.
   *
   * @since 1.3.2
   *
   * @alias cancel
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/cancel-subscription
   *
   * @public ✓ This method is part of the public API
   */
  public delete: CustomersSubscriptionsResource['cancel'] = this.cancel;
  /**
   * List the Customer's Subscriptions.
   *
   * @since 1.3.2
   *
   * @alias list
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/list-subscriptions
   *
   * @public ✓ This method is part of the public API
   */
  public all: CustomersSubscriptionsResource['list'] = this.list;
  /**
   * List the Customer's Subscriptions.
   *
   * @since 3.0.0
   *
   * @alias list
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/list-subscriptions
   *
   * @public ✓ This method is part of the public API
   */
  public page: CustomersSubscriptionsResource['list'] = this.list;

  /**
   * Create a customer subscription.
   *
   * @param params - Create Subscription parameters
   *
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns Customer Subscription
   *
   * @since 1.3.2
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/create-subscription
   *
   * @public ✓ This method is part of the public API
   */
  public create(parameters: CreateParameters): Promise<Subscription>;
  public create(parameters: CreateParameters, callback: Callback<Subscription>): void;
  public create(parameters: CreateParameters) {
    if (renege(this, this.create, ...arguments)) return;
    const customerId = this.getParentId(parameters.customerId);
    if (!checkId(customerId, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    const { customerId: _, ...data } = parameters;
    return this.network.post(this.getResourceUrl(customerId), data);
  }

  /**
   * Get a customer subscription.
   *
   * @param id - Subscription ID
   * @param params - Get Subscription parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @return Customer Subscription
   *
   * @since 1.3.2
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/get-subscription
   *
   * @public ✓ This method is part of the public API
   */
  public get(id: string, parameters: GetParameters): Promise<Subscription>;
  public get(id: string, parameters: GetParameters, callback: Callback<Subscription>): void;
  public get(id: string, parameters: GetParameters) {
    if (renege(this, this.get, ...arguments)) return;
    if (!checkId(id, 'subscription')) {
      throw new ApiError('The subscription id is invalid');
    }
    // parameters || {} is used here, because in case withParent is used, parameters could be omitted.
    const customerId = this.getParentId((parameters || {}).customerId);
    if (!checkId(customerId, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    const { customerId: _, ...query } = parameters || {};
    return this.network.get(`${this.getResourceUrl(customerId)}/${id}`, query);
  }

  /**
   * Get all customer's subscriptions.
   *
   * @param params - List customer's subscriptions parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns A list of found subscriptions
   *
   * @since 3.0.0
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/list-subscriptions
   *
   * @public ✓ This method is part of the public API
   */
  public list(parameters: ListParameters): Promise<List<Subscription>>;
  public list(parameters: ListParameters, callback: Callback<List<Subscription>>): void;
  public list(parameters: ListParameters) {
    if (renege(this, this.list, ...arguments)) return;
    // parameters || {} is used here, because in case withParent is used, parameters could be omitted.
    const customerId = this.getParentId((parameters || {}).customerId);
    if (!checkId(customerId, 'customer')) {
      throw new ApiError('The customer id is invalid');
    }
    const { customerId: _, ...query } = parameters || {};
    return this.network.list(this.getResourceUrl(customerId), 'subscriptions', query).then(result => this.injectPaginationHelpers(result, this.list, parameters || {}));
  }

  /**
   * Update a customer's subscription.
   *
   * @param id - Subscription ID
   * @param params - Update customer subscription parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns The updated Customer Subscription object
   *
   * @since 2.0.0
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/update-subscription
   *
   * @public ✓ This method is part of the public API
   */
  public update(id: string, parameters: UpdateParameters): Promise<Subscription>;
  public update(id: string, parameters: UpdateParameters, callback: Callback<Subscription>): void;
  public update(id: string, parameters: UpdateParameters) {
    if (renege(this, this.update, ...arguments)) return;
    if (!checkId(id, 'subscription')) {
      throw new ApiError('The subscription id is invalid');
    }
    const customerId = this.getParentId(parameters.customerId);
    if (!checkId(customerId, 'customer')) {
      throw new ApiError('The customer is invalid');
    }
    const { customerId: _, ...data } = parameters;
    return this.network.patch(`${this.getResourceUrl(customerId)}/${id}`, data);
  }

  /**
   * Cancel a Subscription
   *
   * @param id - Subscription ID
   * @param params - Delete Subscription parameters
   *                 (DEPRECATED SINCE 3.0.0) Can also be a callback function
   * @param cb - (DEPRECATED SINCE 3.0.0) Callback function, can be used instead of the returned `Promise` object
   *
   * @returns Success status
   *
   * @since 1.3.2
   *
   * @see https://docs.mollie.com/reference/v2/subscriptions-api/cancel-subscription
   *
   * @public ✓ This method is part of the public API
   */
  public cancel(id: string, parameters: CancelParameters): Promise<Subscription>;
  public cancel(id: string, parameters: CancelParameters, callback: Callback<Subscription>): void;
  public cancel(id: string, parameters: CancelParameters) {
    if (renege(this, this.cancel, ...arguments)) return;
    if (!checkId(id, 'subscription')) {
      throw new ApiError('The subscription id is invalid');
    }
    // parameters || {} is used here, because in case withParent is used, parameters could be omitted.
    const customerId = this.getParentId((parameters || {}).customerId);
    if (!checkId(customerId, 'customer')) {
      throw new ApiError('The customer is invalid');
    }
    const { customerId: _, ...query } = parameters || {};
    return this.network.delete<Subscription>(`${this.getResourceUrl(customerId)}/${id}`);
  }
}
