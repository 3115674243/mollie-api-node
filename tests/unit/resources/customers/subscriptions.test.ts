import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import CustomersSubscriptionsResource from '@resources/customers/subscriptions';
import Subscription from '@models/Subscription';
import { MandateMethod } from '@root/types/mandate';
import response from '@tests/unit/__stubs__/customers_subscriptions.json';

const mock = new MockAdapter(axios);

const props = {
  id: 'sub_rVKGtNd6s3',
  customerId: 'cst_stTC2WHAuS',
  amount: {
    currency: 'EUR',
    value: '25.00',
  },
  times: 4,
  interval: '3 months',
  description: 'Quarterly payment',
  method: MandateMethod.creditcard,
  webhookUrl: 'https://example.org/payments/webhook',
};

describe('customers_subscriptions', () => {
  let customersSubscriptions: CustomersSubscriptionsResource;
  beforeEach(() => {
    customersSubscriptions = new CustomersSubscriptionsResource(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(CustomersSubscriptionsResource.resource).toBe('customers_subscriptions');
    expect(CustomersSubscriptionsResource.model).toBe(Subscription);
  });

  describe('.create()', () => {
    mock.onPost(`/customers/${props.customerId}/subscriptions`).reply(200, response._embedded.subscriptions[0]);

    it('should return a subscription instance', () =>
      customersSubscriptions.create(props).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));
  });

  describe('.get()', () => {
    const error = {
      error: { detail: 'The customers_subscription id is invalid' },
    };

    mock.onGet(`/customers/${props.customerId}/subscriptions/${props.id}`).reply(200, response._embedded.subscriptions[0]);
    mock.onGet(`/customers/${props.customerId}/subscriptions/foo`).reply(500, error);

    it('should return a subscription instance', () =>
      customersSubscriptions.get(props.id, { customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));
  });

  describe('.list()', () => {
    mock.onGet(`/customers/${props.customerId}/subscriptions`).reply(200, response);

    it('should return a list of all customer subscriptions', () =>
      customersSubscriptions.all({ customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersSubscriptions
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .all((err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        });
    });

    it('should work with a Promise and with .withParent()', done => {
      customersSubscriptions
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .all()
        .then(result => {
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        })
        .catch(err => {
          expect(err).toBeNull();
          done();
        });
    });
  });

  describe('.cancel()', () => {
    mock.onDelete(`/customers/${props.customerId}/subscriptions/${props.id}`).reply(200, response._embedded.subscriptions[0]);

    it('should return a subscription instance when successfully canceled', () =>
      customersSubscriptions.delete(props.id, { customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersSubscriptions
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .cancel(props.id, (err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Subscription);
          expect(result).toMatchSnapshot();
          done();
        });
    });

    it('should return the status when it could not be canceled', () => {
      customersSubscriptions.delete('alreadycanceled', { customerId: props.customerId }).then(result => {
        expect(result).toEqual(false);
      });
    });
  });
});
