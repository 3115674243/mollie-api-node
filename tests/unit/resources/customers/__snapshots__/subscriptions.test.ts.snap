// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers_subscriptions .cancel() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "mandateId": null,
  "metadata": null,
  "method": null,
  "mode": "live",
  "nextPaymentDate": null,
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "timesRemaining": null,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .cancel() should work with a callback 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "mandateId": null,
  "metadata": null,
  "method": null,
  "mode": "live",
  "nextPaymentDate": null,
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "timesRemaining": null,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .create() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "mandateId": null,
  "metadata": null,
  "method": null,
  "mode": "live",
  "nextPaymentDate": null,
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "timesRemaining": null,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .get() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "mandateId": null,
  "metadata": null,
  "method": null,
  "mode": "live",
  "nextPaymentDate": null,
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "timesRemaining": null,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .list() should return a list of all customer subscriptions 1`] = `
Array [
  Subscription {
    "_links": Object {
      "customer": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "25.00",
    },
    "canceledAt": null,
    "createdAt": "2018-06-01T12:23:34+00:00",
    "description": "Quarterly payment",
    "id": "sub_rVKGtNd6s3",
    "interval": "3 months",
    "mandateId": null,
    "metadata": null,
    "method": null,
    "mode": "live",
    "nextPaymentDate": null,
    "resource": "subscription",
    "startDate": null,
    "status": "active",
    "times": 4,
    "timesRemaining": null,
    "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
  },
]
`;

exports[`customers_subscriptions .list() should work with a Promise and with .withParent() 1`] = `
Array [
  Subscription {
    "_links": Object {
      "customer": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "25.00",
    },
    "canceledAt": null,
    "createdAt": "2018-06-01T12:23:34+00:00",
    "description": "Quarterly payment",
    "id": "sub_rVKGtNd6s3",
    "interval": "3 months",
    "mandateId": null,
    "metadata": null,
    "method": null,
    "mode": "live",
    "nextPaymentDate": null,
    "resource": "subscription",
    "startDate": null,
    "status": "active",
    "times": 4,
    "timesRemaining": null,
    "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
  },
]
`;

exports[`customers_subscriptions .list() should work with a callback 1`] = `
Array [
  Subscription {
    "_links": Object {
      "customer": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "25.00",
    },
    "canceledAt": null,
    "createdAt": "2018-06-01T12:23:34+00:00",
    "description": "Quarterly payment",
    "id": "sub_rVKGtNd6s3",
    "interval": "3 months",
    "mandateId": null,
    "metadata": null,
    "method": null,
    "mode": "live",
    "nextPaymentDate": null,
    "resource": "subscription",
    "startDate": null,
    "status": "active",
    "times": 4,
    "timesRemaining": null,
    "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
  },
]
`;
